pydm/widgets/base.py

```diff
@@ -680,6 +680,12 @@ def value_changed(self, new_val):
         """
         self.value = new_val
         self.channeltype = type(self.value)
+        for base_type in [int, float, str, np.ndarray]:
+            if self.channeltype != base_type and issubclass(self.channeltype, base_type):
+                # Leave bool and int separate, we're concerned about user-defined subclasses, not native Python ones
+                if self.channeltype != bool:
+                    self.channeltype = base_type
+                break
```

Comment:
Not sure how often code gets called, but wouldn't it be more efficient to test for bool once and then in the loop it can just test if something is a subtype.  

``` 
     if self.channeltype !=bool:
         for base_type in [int, float, str, np.ndarray]:
              if issubclass(self.channeltype, base_type):
                 self.channeltype = base_type
              break
```

Notice the check for self.channeltype!=base_type is unnecessary.  If it is ``int`` then it would pass the subclass check and set channeltype to ``int`` which it already was and then break.

Reply:
I agree with @Thrameos's sentiment here.  `value_changed` gets called _a lot_, maybe more than any other method in all of PyDM, so it pays to make this check as lightweight as possible.

Maybe something like this?
```
if self.channeltype not in (int, float, bool, str, np.ndarray):
    for base_type in (int, float, str, np.ndarray):
        if issubclass(self.channeltype, base_type): 
            self.channeltype = base_type
        break
```

Reply:
If this is a critical section then skipping the ``for`` loop and the implicit loop from "not in" entirely would be the best policy.  Instead use a global table cache which would be O(1).

```
         value_type = type(value)
         self.channeltype = TYPE_TABLE.get(value_type, None)
         if self.channeltype is None:
             # assume we will use the type as is
             self.channeltype = value_type

             # check for user defined subtypes
             for base_type in (int, float, str, np.ndarray):
                if issubclass(self.channeltype, base_type): 
                   self.channeltype = base_type
                   break

             # Cache the result so all future look ups are O(1)
             TABLE_TYPE[value_type] = self.channeltype
```

The table would be prepopulated with ``TYPE_TABLE[bool] = bool``

Each time it is called we have one hash map look up and a comparison.  It it is not found then the search occurs to find the best type.   That result is cached so it never has to evaluate the loop for that type again.